File: backend/src/main/java/com/codeloom/backend/controller/ProblemController.java
package com.codeloom.backend.controller;

import com.codeloom.backend.dto.CreateProblemRequest;
import com.codeloom.backend.dto.ProblemDto;
import com.codeloom.backend.dto.ProblemFilters;
import com.codeloom.backend.dto.ProblemListDto;
import com.codeloom.backend.dto.UpdateProblemRequest;
import com.codeloom.backend.model.Problem;
import com.codeloom.backend.model.ProblemDifficulty;
import com.codeloom.backend.service.ProblemService;
import jakarta.validation.Valid;
import java.util.List;
import java.util.Set;
import lombok.RequiredArgsConstructor;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/v1/problems")
@RequiredArgsConstructor
public class ProblemController {
    private final ProblemService service;

    @GetMapping("items")
    public List<ProblemListDto> findAllItems(
            Authentication authentication,
            @RequestParam(required = false) Set<ProblemDifficulty> difficulties,
            @RequestParam(required = false) Set<String> topics,
            @RequestParam(defaultValue = "true") boolean publishedOnly) {
        var filters = ProblemFilters.builder()
                .difficulties(difficulties)
                .publishedOnly(publishedOnly)
                .topics(topics)
                .build();
        return service.findItemsByFilters(authentication, filters);
    }

    @GetMapping("{problemId}")
    public Problem findById(Authentication authentication, @PathVariable long problemId) {
        return service.findById(authentication, problemId);
    }

    @GetMapping("slug/{problemSlug}")
    public ProblemDto findDtoBySlug(Authentication authentication, @PathVariable String problemSlug) {
        return service.findDtoBySlug(authentication, problemSlug);
    }

    @DeleteMapping("{problemId}")
    public void delete(@PathVariable long problemId) {
        service.deleteById(problemId);
    }

    @PostMapping
    public Problem create(@Valid @RequestBody CreateProblemRequest request) {
        return service.create(request);
    }

    @PutMapping("/{problemId}")
    public Problem update(@PathVariable long problemId, @Valid @RequestBody UpdateProblemRequest request) {
        return service.update(problemId, request);
    }

    @PatchMapping("/{problemId}/publish")
    public void publish(@PathVariable long problemId) {
        service.publish(problemId);
    }

    @PatchMapping("/{problemId}/unpublish")
    public void unpublish(@PathVariable long problemId) {
        service.unpublish(problemId);
    }
}

File: backend/src/main/kotlin/com/codeloom/backend/controller/ProblemController.kt
package com.codeloom.backend.controller

import com.codeloom.backend.config.BAD_REQUEST_RESPONSE_REF
import com.codeloom.backend.config.FORBIDDEN_RESPONSE_REF
import com.codeloom.backend.config.NOT_FOUND_RESPONSE_REF
import com.codeloom.backend.dto.CreateProblemRequest
import com.codeloom.backend.dto.ProblemDto
import com.codeloom.backend.dto.ProblemFilters
import com.codeloom.backend.dto.ProblemListDto
import com.codeloom.backend.model.Problem
import com.codeloom.backend.model.ProblemDifficulty
import com.codeloom.backend.service.ProblemService
import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.responses.ApiResponse
import io.swagger.v3.oas.annotations.tags.Tag
import jakarta.validation.Valid
import org.springframework.security.core.Authentication
import org.springframework.web.bind.annotation.*
import tools.jackson.databind.JsonNode

@RestController
@RequestMapping("/v1/problems")
@Tag(name = "Problems API")
class ProblemController(
    private val problemService: ProblemService,
) {
    @Operation(summary = "Get all problem items")
    @ApiResponse(responseCode = "403", ref = FORBIDDEN_RESPONSE_REF)
    @GetMapping("items")
    fun findAllItems(
        auth: Authentication,
        @RequestParam(required = false) difficulties: Set<ProblemDifficulty>?,
        @RequestParam(required = false) topics: Set<String>?,
        @RequestParam(defaultValue = "true") publishedOnly: Boolean,
    ): List<ProblemListDto> {
        return problemService.findItemsByFilters(
            auth,
            ProblemFilters(
                difficulties = difficulties,
                topics = topics,
                publishedOnly = publishedOnly,
            ),
        )
    }

    @Operation(summary = "Get problem by id")
    @ApiResponse(responseCode = "404", ref = NOT_FOUND_RESPONSE_REF)
    @GetMapping("{problemId}")
    fun findById(
        auth: Authentication,
        @PathVariable("problemId") problemId: Long,
    ): Problem {
        return problemService.findById(auth, problemId)
    }

    @Operation(summary = "Get problem dto by slug")
    @ApiResponse(responseCode = "404", ref = NOT_FOUND_RESPONSE_REF)
    @GetMapping("slug/{problemSlug}")
    fun findDtoBySlug(
        auth: Authentication,
        @PathVariable("problemSlug") problemSlug: String,
    ): ProblemDto = problemService.findDtoBySlug(auth, problemSlug)

    @Operation(summary = "Delete problem by id")
    @DeleteMapping("{problemId}")
    fun delete(
        @PathVariable("problemId") problemId: Long,
    ) {
        problemService.deleteById(problemId)
    }

    @Operation(summary = "Create new problem")
    @ApiResponse(responseCode = "400", ref = BAD_REQUEST_RESPONSE_REF)
    @PostMapping
    fun create(
        @Valid @RequestBody request: CreateProblemRequest,
    ): Problem {
        return problemService.create(request)
    }

    @Operation(summary = "Update problem")
    @ApiResponse(responseCode = "400", ref = BAD_REQUEST_RESPONSE_REF)
    @ApiResponse(responseCode = "404", ref = NOT_FOUND_RESPONSE_REF)
    @PutMapping("/{problemId}")
    fun update(
        @PathVariable problemId: Long,
        @RequestBody updated: JsonNode,
    ): Problem {
        return problemService.update(problemId, updated)
    }

    @Operation(summary = "Publish problem")
    @ApiResponse(responseCode = "404", ref = NOT_FOUND_RESPONSE_REF)
    @PatchMapping("/{problemId}/publish")
    fun publish(
        @PathVariable("problemId") problemId: Long,
    ) = problemService.publish(problemId)

    @Operation(summary = "Unpublish problem")
    @ApiResponse(responseCode = "404", ref = NOT_FOUND_RESPONSE_REF)
    @PatchMapping("/{problemId}/unpublish")
    fun unpublish(
        @PathVariable("problemId") problemId: Long,
    ) = problemService.unpublish(problemId)
}
